File: permutation.rb
# k elements in n permutation
def permute(nums, k)
  @result, @nums, @k = [], nums, k
  dfs []
  @result
end

def dfs(arr)
  @result << arr if arr.size == @k
  (@nums - arr).each { |n| dfs arr + [n] }
end

p permute [1, 2, 3, 4], 3

File: permutation.go
package main

import (
	"fmt"
)

func permutation(arr []int, size int) {
	dfs(arr, []int{}, size)
}

func dfs(remain []int, current []int, size int) {
	if len(current) == size {
		fmt.Println(current)
	}
	newRemain := []int{}
	newCurrent := current

	for i, v := range remain {
		newRemain = []int{}
		newRemain = append(newRemain, remain[:i]...)
		newRemain = append(newRemain, remain[i+1:]...)
		dfs(newRemain, append(newCurrent, v), size)
	}
}

func main() {
	arr := []int{1, 2, 3, 4}
	permutation(arr, 4)
}
